File: examples/go/hand/agg/handlers/game_rules.go
// Package handlers provides poker game rules and hand evaluation.
package handlers

import (
	"sort"

	"github.com/benjaminabbitt/angzarr/client/go/proto/examples"
)

// HandRank represents a hand ranking result with score for comparison.
type HandRank struct {
	RankType examples.HandRankType
	Score    int32
	Kickers  []examples.Rank
}

// GameRules defines variant-specific poker logic.
type GameRules interface {
	// HoleCardCount returns the number of hole cards dealt to each player.
	HoleCardCount() int

	// EvaluateHand evaluates a player's hand against community cards.
	EvaluateHand(holeCards, communityCards []*examples.Card) *HandRank

	// UsesCommunityCards returns true if the variant uses community cards.
	UsesCommunityCards() bool
}

// GetRules returns the game rules for a variant.
func GetRules(variant examples.GameVariant) GameRules {
	switch variant {
	case examples.GameVariant_TEXAS_HOLDEM:
		return &TexasHoldemRules{}
	case examples.GameVariant_OMAHA:
		return &OmahaRules{}
	case examples.GameVariant_FIVE_CARD_DRAW:
		return &FiveCardDrawRules{}
	default:
		return &TexasHoldemRules{}
	}
}

// TexasHoldemRules implements Texas Hold'em rules.
type TexasHoldemRules struct{}

func (r *TexasHoldemRules) HoleCardCount() int { return 2 }

func (r *TexasHoldemRules) EvaluateHand(holeCards, communityCards []*examples.Card) *HandRank {
	allCards := append(holeCards, communityCards...)
	if len(allCards) < 5 {
		return &HandRank{RankType: examples.HandRankType_HIGH_CARD, Score: 0}
	}
	return findBestFive(allCards)
}

func (r *TexasHoldemRules) UsesCommunityCards() bool { return true }

// OmahaRules implements Omaha rules (must use exactly 2 hole + 3 community).
type OmahaRules struct{}

func (r *OmahaRules) HoleCardCount() int { return 4 }

func (r *OmahaRules) EvaluateHand(holeCards, communityCards []*examples.Card) *HandRank {
	if len(holeCards) < 2 || len(communityCards) < 3 {
		return &HandRank{RankType: examples.HandRankType_HIGH_CARD, Score: 0}
	}

	var bestRank *HandRank

	// Must use exactly 2 from hole, 3 from community
	holeCombos := combinations(holeCards, 2)
	commCombos := combinations(communityCards, 3)

	for _, holeCombo := range holeCombos {
		for _, commCombo := range commCombos {
			five := append(holeCombo, commCombo...)
			rank := evaluateFive(five)
			if bestRank == nil || rank.Score > bestRank.Score {
				bestRank = rank
			}
		}
	}

	if bestRank == nil {
		return &HandRank{RankType: examples.HandRankType_HIGH_CARD, Score: 0}
	}
	return bestRank
}

func (r *OmahaRules) UsesCommunityCards() bool { return true }

// FiveCardDrawRules implements Five Card Draw rules.
type FiveCardDrawRules struct{}

func (r *FiveCardDrawRules) HoleCardCount() int { return 5 }

func (r *FiveCardDrawRules) EvaluateHand(holeCards, _ []*examples.Card) *HandRank {
	if len(holeCards) < 5 {
		return &HandRank{RankType: examples.HandRankType_HIGH_CARD, Score: 0}
	}
	return evaluateFive(holeCards[:5])
}

func (r *FiveCardDrawRules) UsesCommunityCards() bool { return false }

// findBestFive finds the best 5-card hand from any number of cards.
func findBestFive(cards []*examples.Card) *HandRank {
	if len(cards) < 5 {
		return &HandRank{RankType: examples.HandRankType_HIGH_CARD, Score: 0}
	}

	var bestRank *HandRank

	combos := combinations(cards, 5)
	for _, combo := range combos {
		rank := evaluateFive(combo)
		if bestRank == nil || rank.Score > bestRank.Score {
			bestRank = rank
		}
	}

	if bestRank == nil {
		return &HandRank{RankType: examples.HandRankType_HIGH_CARD, Score: 0}
	}
	return bestRank
}

// evaluateFive evaluates exactly 5 cards.
func evaluateFive(cards []*examples.Card) *HandRank {
	if len(cards) != 5 {
		return &HandRank{RankType: examples.HandRankType_HIGH_CARD, Score: 0}
	}

	// Count ranks and suits
	rankCounts := make(map[int32]int)
	suitCounts := make(map[int32]int)
	ranks := make([]int32, 5)

	for i, card := range cards {
		rankCounts[int32(card.Rank)]++
		suitCounts[int32(card.Suit)]++
		ranks[i] = int32(card.Rank)
	}

	// Sort ranks descending
	sort.Slice(ranks, func(i, j int) bool { return ranks[i] > ranks[j] })

	isFlush := false
	for _, count := range suitCounts {
		if count == 5 {
			isFlush = true
			break
		}
	}

	isStraight := checkStraight(ranks)
	isWheel := checkWheel(ranks)

	// Build count pattern for pair/trip/quad detection
	type rankCount struct {
		rank  int32
		count int
	}
	var counts []rankCount
	for rank, count := range rankCounts {
		counts = append(counts, rankCount{rank, count})
	}
	// Sort by count desc, then by rank desc
	sort.Slice(counts, func(i, j int) bool {
		if counts[i].count != counts[j].count {
			return counts[i].count > counts[j].count
		}
		return counts[i].rank > counts[j].rank
	})

	countPattern := make([]int, len(counts))
	for i, c := range counts {
		countPattern[i] = c.count
	}

	// Determine hand type and score
	if isFlush && (isStraight || isWheel) {
		high := ranks[0]
		if isWheel {
			high = 5 // Wheel straight flush high is 5
		}
		if high == int32(examples.Rank_ACE) && !isWheel {
			// Royal flush
			return &HandRank{
				RankType: examples.HandRankType_ROYAL_FLUSH,
				Score:    10_000_000,
				Kickers:  nil,
			}
		}
		return &HandRank{
			RankType: examples.HandRankType_STRAIGHT_FLUSH,
			Score:    9_000_000 + high,
			Kickers:  []examples.Rank{examples.Rank(high)},
		}
	}

	if len(countPattern) >= 2 && countPattern[0] == 4 && countPattern[1] == 1 {
		quadRank := counts[0].rank
		kicker := counts[1].rank
		return &HandRank{
			RankType: examples.HandRankType_FOUR_OF_A_KIND,
			Score:    8_000_000 + quadRank*100 + kicker,
			Kickers:  []examples.Rank{examples.Rank(quadRank), examples.Rank(kicker)},
		}
	}

	if len(countPattern) >= 2 && countPattern[0] == 3 && countPattern[1] == 2 {
		tripsRank := counts[0].rank
		pairRank := counts[1].rank
		return &HandRank{
			RankType: examples.HandRankType_FULL_HOUSE,
			Score:    7_000_000 + tripsRank*100 + pairRank,
			Kickers:  []examples.Rank{examples.Rank(tripsRank), examples.Rank(pairRank)},
		}
	}

	if isFlush {
		score := rankScore(ranks)
		kickers := make([]examples.Rank, len(ranks))
		for i, r := range ranks {
			kickers[i] = examples.Rank(r)
		}
		return &HandRank{
			RankType: examples.HandRankType_FLUSH,
			Score:    6_000_000 + score,
			Kickers:  kickers,
		}
	}

	if isStraight || isWheel {
		high := ranks[0]
		if isWheel {
			high = 5
		}
		return &HandRank{
			RankType: examples.HandRankType_STRAIGHT,
			Score:    5_000_000 + high,
			Kickers:  []examples.Rank{examples.Rank(high)},
		}
	}

	if len(countPattern) >= 3 && countPattern[0] == 3 && countPattern[1] == 1 && countPattern[2] == 1 {
		tripsRank := counts[0].rank
		kicker1 := counts[1].rank
		kicker2 := counts[2].rank
		kickerScore := kicker1*50 + kicker2
		return &HandRank{
			RankType: examples.HandRankType_THREE_OF_A_KIND,
			Score:    4_000_000 + tripsRank*1000 + kickerScore,
			Kickers:  []examples.Rank{examples.Rank(tripsRank), examples.Rank(kicker1), examples.Rank(kicker2)},
		}
	}

	if len(countPattern) >= 3 && countPattern[0] == 2 && countPattern[1] == 2 && countPattern[2] == 1 {
		highPair := counts[0].rank
		lowPair := counts[1].rank
		kicker := counts[2].rank
		return &HandRank{
			RankType: examples.HandRankType_TWO_PAIR,
			Score:    3_000_000 + highPair*1000 + lowPair*50 + kicker,
			Kickers:  []examples.Rank{examples.Rank(highPair), examples.Rank(lowPair), examples.Rank(kicker)},
		}
	}

	if len(countPattern) >= 4 && countPattern[0] == 2 && countPattern[1] == 1 && countPattern[2] == 1 && countPattern[3] == 1 {
		pairRank := counts[0].rank
		kicker1 := counts[1].rank
		kicker2 := counts[2].rank
		kicker3 := counts[3].rank
		kickerScore := kicker1*1000 + kicker2*50 + kicker3
		return &HandRank{
			RankType: examples.HandRankType_PAIR,
			Score:    2_000_000 + pairRank*10000 + kickerScore,
			Kickers:  []examples.Rank{examples.Rank(pairRank), examples.Rank(kicker1), examples.Rank(kicker2), examples.Rank(kicker3)},
		}
	}

	// High card
	score := rankScore(ranks)
	kickers := make([]examples.Rank, len(ranks))
	for i, r := range ranks {
		kickers[i] = examples.Rank(r)
	}
	return &HandRank{
		RankType: examples.HandRankType_HIGH_CARD,
		Score:    1_000_000 + score,
		Kickers:  kickers,
	}
}

// checkStraight checks if sorted ranks (descending) form a straight.
func checkStraight(ranks []int32) bool {
	if len(ranks) != 5 {
		return false
	}
	for i := 0; i < 4; i++ {
		if ranks[i]-ranks[i+1] != 1 {
			return false
		}
	}
	return true
}

// checkWheel checks for wheel straight (A-2-3-4-5).
func checkWheel(ranks []int32) bool {
	sorted := make([]int32, len(ranks))
	copy(sorted, ranks)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i] < sorted[j] })

	// A-2-3-4-5 means sorted is [2, 3, 4, 5, 14]
	return len(sorted) == 5 &&
		sorted[0] == 2 &&
		sorted[1] == 3 &&
		sorted[2] == 4 &&
		sorted[3] == 5 &&
		sorted[4] == int32(examples.Rank_ACE)
}

// rankScore calculates rank-based score for high card comparison.
func rankScore(ranks []int32) int32 {
	var score int32
	multiplier := int32(10000)
	for i := 0; i < 5 && i < len(ranks); i++ {
		score += ranks[i] * multiplier
		multiplier /= 15 // Ensure no overlap
	}
	return score
}

// combinations generates all k-combinations of cards.
func combinations(cards []*examples.Card, k int) [][]*examples.Card {
	if k > len(cards) {
		return nil
	}
	if k == 0 {
		return [][]*examples.Card{{}}
	}
	if k == len(cards) {
		result := make([]*examples.Card, len(cards))
		copy(result, cards)
		return [][]*examples.Card{result}
	}

	var result [][]*examples.Card
	var helper func(start int, current []*examples.Card)
	helper = func(start int, current []*examples.Card) {
		if len(current) == k {
			combo := make([]*examples.Card, k)
			copy(combo, current)
			result = append(result, combo)
			return
		}
		for i := start; i < len(cards); i++ {
			helper(i+1, append(current, cards[i]))
		}
	}
	helper(0, nil)
	return result
}

File: examples/python/hand/agg/handlers/game_rules.py
"""Polymorphic game rules for different poker variants."""

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Optional
import random

from angzarr_client.proto.examples import poker_types_pb2 as poker_types


@dataclass
class DealResult:
    """Result of dealing cards."""

    player_cards: dict  # player_root -> list of (suit, rank) tuples
    remaining_deck: list  # Remaining cards in deck


@dataclass
class DrawResult:
    """Result of executing a draw."""

    new_hole_cards: list  # Updated hole cards after draw
    cards_drawn: list  # New cards drawn from deck
    remaining_deck: list  # Remaining cards in deck


@dataclass
class PhaseTransition:
    """Information about phase transition."""

    next_phase: int  # BettingPhase enum
    community_cards_to_deal: int
    is_showdown: bool


class GameRules(ABC):
    """Abstract base class for poker variant rules."""

    @property
    @abstractmethod
    def variant(self) -> int:
        """Return the GameVariant enum value."""
        pass

    @property
    @abstractmethod
    def hole_card_count(self) -> int:
        """Number of hole cards dealt to each player."""
        pass

    @property
    @abstractmethod
    def phases(self) -> list:
        """List of betting phases for this variant."""
        pass

    @abstractmethod
    def get_next_phase(self, current_phase: int) -> Optional[PhaseTransition]:
        """Get the next phase after the current one, or None if hand is complete."""
        pass

    @abstractmethod
    def evaluate_hand(self, hole_cards: list, community_cards: list) -> tuple:
        """
        Evaluate a hand and return (rank_type, score, kickers).

        rank_type: HandRankType enum
        score: Numeric score for comparison
        kickers: List of Rank values for tie-breaking
        """
        pass

    def create_deck(self, seed: Optional[bytes] = None) -> list:
        """Create and shuffle a standard 52-card deck."""
        deck = []
        for suit in [
            poker_types.CLUBS,
            poker_types.DIAMONDS,
            poker_types.HEARTS,
            poker_types.SPADES,
        ]:
            for rank in range(2, 15):  # 2-14 (Ace)
                deck.append((suit, rank))

        if seed:
            rng = random.Random(int.from_bytes(seed[:8], "big"))
            rng.shuffle(deck)
        else:
            random.shuffle(deck)

        return deck

    def deal_hole_cards(
        self, deck: list, players: list, seed: Optional[bytes] = None
    ) -> DealResult:
        """Deal hole cards to all players."""
        if seed:
            working_deck = self.create_deck(seed)
        elif not deck:
            working_deck = self.create_deck()
        else:
            working_deck = deck.copy()

        player_cards = {}
        for player_root in players:
            cards = []
            for _ in range(self.hole_card_count):
                if working_deck:
                    cards.append(working_deck.pop())
            player_cards[player_root] = cards

        return DealResult(
            player_cards=player_cards,
            remaining_deck=working_deck,
        )


class TexasHoldemRules(GameRules):
    """Rules for Texas Hold'em poker."""

    @property
    def variant(self) -> int:
        return poker_types.TEXAS_HOLDEM

    @property
    def hole_card_count(self) -> int:
        return 2

    @property
    def phases(self) -> list:
        return [
            poker_types.PREFLOP,
            poker_types.FLOP,
            poker_types.TURN,
            poker_types.RIVER,
            poker_types.SHOWDOWN,
        ]

    def get_next_phase(self, current_phase: int) -> Optional[PhaseTransition]:
        if current_phase == poker_types.PREFLOP:
            return PhaseTransition(
                next_phase=poker_types.FLOP,
                community_cards_to_deal=3,
                is_showdown=False,
            )
        elif current_phase == poker_types.FLOP:
            return PhaseTransition(
                next_phase=poker_types.TURN,
                community_cards_to_deal=1,
                is_showdown=False,
            )
        elif current_phase == poker_types.TURN:
            return PhaseTransition(
                next_phase=poker_types.RIVER,
                community_cards_to_deal=1,
                is_showdown=False,
            )
        elif current_phase == poker_types.RIVER:
            return PhaseTransition(
                next_phase=poker_types.SHOWDOWN,
                community_cards_to_deal=0,
                is_showdown=True,
            )
        return None

    def evaluate_hand(self, hole_cards: list, community_cards: list) -> tuple:
        """Evaluate best 5-card hand from 7 cards (2 hole + 5 community)."""
        all_cards = hole_cards + community_cards
        return self._find_best_hand(all_cards)

    def _find_best_hand(self, cards: list) -> tuple:
        """Find the best 5-card hand from available cards."""
        if len(cards) < 5:
            return (poker_types.HIGH_CARD, 0, [])

        from itertools import combinations

        best = (poker_types.HIGH_CARD, 0, [])
        for combo in combinations(cards, 5):
            result = self._evaluate_five(list(combo))
            if result[1] > best[1]:
                best = result

        return best

    def _evaluate_five(self, cards: list) -> tuple:
        """Evaluate exactly 5 cards."""
        suits = [c[0] for c in cards]
        ranks = sorted([c[1] for c in cards], reverse=True)

        is_flush = len(set(suits)) == 1
        is_straight = self._is_straight(ranks)

        rank_counts = {}
        for r in ranks:
            rank_counts[r] = rank_counts.get(r, 0) + 1

        counts = sorted(rank_counts.values(), reverse=True)
        sorted_by_count = sorted(
            rank_counts.keys(), key=lambda x: (rank_counts[x], x), reverse=True
        )

        # Calculate score
        if is_straight and is_flush:
            if ranks == [14, 13, 12, 11, 10]:
                return (poker_types.ROYAL_FLUSH, 10000000, [])
            return (poker_types.STRAIGHT_FLUSH, 9000000 + ranks[0], [])
        elif counts == [4, 1]:
            return (
                poker_types.FOUR_OF_A_KIND,
                8000000 + sorted_by_count[0] * 100,
                [sorted_by_count[1]],
            )
        elif counts == [3, 2]:
            return (
                poker_types.FULL_HOUSE,
                7000000 + sorted_by_count[0] * 100 + sorted_by_count[1],
                [],
            )
        elif is_flush:
            return (poker_types.FLUSH, 6000000 + self._rank_score(ranks), ranks)
        elif is_straight:
            return (poker_types.STRAIGHT, 5000000 + ranks[0], [])
        elif counts == [3, 1, 1]:
            kickers = [r for r in sorted_by_count if rank_counts[r] == 1]
            return (
                poker_types.THREE_OF_A_KIND,
                4000000 + sorted_by_count[0] * 1000,
                kickers,
            )
        elif counts == [2, 2, 1]:
            pairs = [r for r in sorted_by_count if rank_counts[r] == 2]
            kicker = [r for r in sorted_by_count if rank_counts[r] == 1]
            return (
                poker_types.TWO_PAIR,
                3000000 + max(pairs) * 100 + min(pairs),
                kicker,
            )
        elif counts == [2, 1, 1, 1]:
            pair = [r for r in sorted_by_count if rank_counts[r] == 2][0]
            kickers = [r for r in sorted_by_count if rank_counts[r] == 1]
            return (poker_types.PAIR, 2000000 + pair * 1000, kickers)
        else:
            return (poker_types.HIGH_CARD, 1000000 + self._rank_score(ranks), ranks)

    def _is_straight(self, ranks: list) -> bool:
        """Check if sorted ranks form a straight."""
        if ranks == [14, 5, 4, 3, 2]:  # Wheel (A-2-3-4-5)
            return True
        for i in range(len(ranks) - 1):
            if ranks[i] - ranks[i + 1] != 1:
                return False
        return True

    def _rank_score(self, ranks: list) -> int:
        """Calculate a score from ranks for comparison."""
        score = 0
        for i, r in enumerate(ranks):
            score += r * (15 ** (4 - i))
        return score


class OmahaRules(TexasHoldemRules):
    """Rules for Omaha poker (uses 2 of 4 hole cards)."""

    @property
    def variant(self) -> int:
        return poker_types.OMAHA

    @property
    def hole_card_count(self) -> int:
        return 4

    def evaluate_hand(self, hole_cards: list, community_cards: list) -> tuple:
        """Evaluate best hand using exactly 2 hole cards and 3 community cards."""
        from itertools import combinations

        best = (poker_types.HIGH_CARD, 0, [])

        # Must use exactly 2 hole cards and 3 community cards
        for hole_combo in combinations(hole_cards, 2):
            for comm_combo in combinations(community_cards, 3):
                cards = list(hole_combo) + list(comm_combo)
                result = self._evaluate_five(cards)
                if result[1] > best[1]:
                    best = result

        return best


class FiveCardDrawRules(GameRules):
    """Rules for Five Card Draw poker."""

    @property
    def variant(self) -> int:
        return poker_types.FIVE_CARD_DRAW

    @property
    def hole_card_count(self) -> int:
        return 5

    @property
    def phases(self) -> list:
        return [
            poker_types.PREFLOP,  # Initial betting
            poker_types.DRAW,  # Draw phase
            poker_types.SHOWDOWN,
        ]

    def get_next_phase(self, current_phase: int) -> Optional[PhaseTransition]:
        if current_phase == poker_types.PREFLOP:
            return PhaseTransition(
                next_phase=poker_types.DRAW,
                community_cards_to_deal=0,
                is_showdown=False,
            )
        elif current_phase == poker_types.DRAW:
            return PhaseTransition(
                next_phase=poker_types.SHOWDOWN,
                community_cards_to_deal=0,
                is_showdown=True,
            )
        return None

    def evaluate_hand(self, hole_cards: list, community_cards: list) -> tuple:
        """Evaluate 5-card hand (no community cards in draw)."""
        # Reuse Hold'em evaluation since it's standard poker hand ranking
        evaluator = TexasHoldemRules()
        return evaluator._evaluate_five(hole_cards)

    def execute_draw(
        self, deck: list, hole_cards: list, discard_indices: list
    ) -> DrawResult:
        """
        Execute a draw - discard selected cards and draw replacements.

        Args:
            deck: Current deck to draw from
            hole_cards: Player's current hole cards
            discard_indices: Indices of cards to discard (0-4)

        Returns:
            DrawResult with new hand, cards drawn, and remaining deck
        """
        new_hole = hole_cards.copy()
        remaining_deck = deck.copy()

        # Remove discarded cards (in reverse order to maintain indices)
        for i in sorted(discard_indices, reverse=True):
            new_hole.pop(i)

        # Draw new cards
        cards_drawn = []
        for _ in range(len(discard_indices)):
            if remaining_deck:
                card = remaining_deck.pop()
                cards_drawn.append(card)
                new_hole.append(card)

        return DrawResult(
            new_hole_cards=new_hole,
            cards_drawn=cards_drawn,
            remaining_deck=remaining_deck,
        )


def get_game_rules(variant: int) -> GameRules:
    """Factory function to get rules for a game variant."""
    if variant == poker_types.TEXAS_HOLDEM:
        return TexasHoldemRules()
    elif variant == poker_types.OMAHA:
        return OmahaRules()
    elif variant == poker_types.FIVE_CARD_DRAW:
        return FiveCardDrawRules()
    else:
        # Default to Hold'em
        return TexasHoldemRules()
